### ISSUE TITLE
port sims (@pixelzoom)

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/tasks#354]
Creating this task to evaluate porting of sims that I've worked on from Java to HTML5.

**Comment by pixelzoom** [comment]
Here are the Java-based PhET simulations that I was primarily responsible for, and their porting status.

**Completed:**
- Acid-Base Solutions (MLL)
- Balancing Chemical Equations (MLL)
- Beer's Law Lab (@pixelzoom)
- Color Vision (@aaronsamuel137)
- Concentration  (@pixelzoom)
- Faraday's Electromagnetic Lab, Generator, Magnet and Compass, Magnets and Electromagnets (@pixelzoom)
- Fourier: Making Waves (@pixelzoom)
- Graphing Lines (@pixelzoom)
- Models of the Hydrogen Atom (@pixelzoom)
- Molarity (@pixelzoom)
- pH Scale  (@pixelzoom)
- pH Scale: Basics  (@pixelzoom)
- Reactants, Products and Leftovers  (@pixelzoom)
- Rutherford Scattering (@jessegreenberg)

**Partial completed:**
- Capacitor Lab (@jessegreenberg, partial port in "Basics" version)
- Molecule Polarity (@pixelzoom, no "Real Molecules" screen)

**Not started**
- Glaciers
- Optical Quantum Control
- Optical Tweezers and Applications, Molecular Motors (2 sims)
- Quantum Bound States, Band Structure, Double Wells and Covalent Bounds (3 sims)
- Quantum Tunneling and Wave Packets

**Comment by pixelzoom** [comment]
From an email to @ariel-phet on 8/25/2014, an overview of various Java "projects" that I worked on that have not yet been ported to HTML5.

**bound-states**
- lines of code: 21522
- Unfuddle issues: 2
- graphics: Piccolo
- model: Listener
- 3rd-party: SchmidtLeeSolver.java (GPL)
- sims: 
  - Quantum Bound States (3 screens)
  - Double Wells and Covalent Bonds (1 screen from QBS)
  - Band Structure (1 screen from QBS)

**capacitor-lab**
- lines of code: 12710
- Unfuddle issues: 0
- graphics: Piccolo
- model: Property + Listener
- sims: 
  - Capacitor Lab (2 screens)

✅ **faraday**
- lines of code: 15351
- Unfuddle issues: 15
- graphics: phetgraphics
- model: Listener
- sims:
  - Faraday’s Electromagnetic Lab (5 screens)
  - Magnet and Compass (1 screen)
  - Magnets and Electromagnets (2 screens)
  - Generator (5 screens)

✅ **fourier**
- lines of code: 19286
- Unfuddle issues: 18
- graphics: phetgraphics
- model: Listener
- sims: 
  - Fourier: Making Waves (3 screens)

**glaciers**
- lines of code: 14966
- Unfuddle issues: 3
- graphics: Piccolo
- model: Listener
- sims: 
  - Glaciers (2 screens)

✅ **hydrogen-atom**
- lines of code: 18357
- Unfuddle issues: 1
- graphics: Piccolo
- model: Listener
- sims: 
  - Models of the Hydrogen Atom (1 screen)

**optical-quantum-control**
- lines of code: 5181
- Unfuddle issues: 1
- graphics: phetgraphics
- model: Listener
- sims: 
  - Optical Quantum Control (1 screen)

**optical-tweezers**
- lines of code: 19985
- Unfuddle issues: 4
- graphics: Piccolo
- model: Listener
- sims:
  - Optical Tweezers (3 screens)
  - Stretching DNA (1 screen)
  - Molecular Motors (1 screen)

**quantum-tunneling**
- lines of code: 18527
- Unfuddle issues: 3
- graphics: Piccolo
- model: Listener
- 3rd-party: SchmidtLeeSolver.java (GPL), RichardsonSolver.java (license?)
- sims: 
  - Quantum Tunneling and Wave Packets (1 screen)

✅ **rutherford-scattering**
- lines of code: 4686
- Unfuddle issues: 2
- graphics: Piccolo
- model: Listener
- sims: 
  - Rutherford Scattering (2 screens)

**Comment by pixelzoom** [comment]
From 2/7/2014 email to @ariel-phet, overview of Java chem sims that had not yet been ported to HTML5.  This includes sims that I didn't implement.

✅ **Model of the Hydrogen Atom**
- Java project: hydrogen-atom
- first published: 1/2007
- original developer: CM
- lines of code: 18K
- Unfuddle tickets: none
- graphics: Piccolo2D
- notification: Listener
- 3D view of deBroglie model relies on wireframe code (~800 lines) from Sun that will require porting
- moving photons is a likely performance issue, ~24 of them visible, will effect responsiveness of wavelength slider
- lots of image files for buttons and controls, cost will increase if you want to do these programmatically to get a specific 'look'
- estimated time to port: 6 weeks

✅ **Rutherford Scattering**
- Java project: rutherford-scattering
- first published: 3/2007
- original developer: CM
- lines of code: 5K
- Unfuddle tickets: none
- graphics: Piccolo2D
- notification: Listener
- animation of alpha particles is a likely performance issue, will affect responsiveness of controls
- estimated time to port: 3 weeks

✅ **Molecule Polarity ("Real Molecules" tab)**
- small amount of code specific to this tab (~500 lines), heavy lifting is done by Jmol
- requires investigation of JSmol
- estimated time to port: ? (I would guess 2-4 weeks)

**Gas Properties, Ballons & Buoyancy, Reversible Reactions**
- Java project: idea-gas
- these 3 sims are part of one Java project, and are tightly coupled, which will dramatically increase cost of port
- first published: unknown, prior to 2005
- original developer: Ron LeMaster
- lines of code:15K
- Unfuddle tickets: many, some that are 'in progress'
- graphics: phetgraphics
- notification: Listener
- hundreds (thousands?) of particles animating with random walk makes feasibility questionable
- estimated time to port: ? (more than 15 weeks for the suite)

**Reactions and Rates**
- Java project: reactions-and-rates
- first published: unknown, prior to 2009
- original developer: Ron LeMaster
- lines of code: 17K
- Unfuddle tickets: one 'severe', one 'TODO' list, several others
- graphics: Piccolo2D
- notification: Listener
- dozens to hundreds of animating molecules
- estimated time to port: ? (more than 10 weeks)

**Neon Lights & Other Discharge Lamps**
- Java project: discharge-lamps
- first published: unknown, prior to 2008
- original developer: Ron LeMaster
- lines of code: 7K
- Unfuddle tickets: a few, should be reviewed
- graphics: Piccolo2D
- notification: Listener
- lots of animated electrons, atoms and photons
- this sim has a dependency on 'lasers' (another sim), which includes another 8K line of code
- estimated time to port: ? (more than 10 weeks)

**Comment by ariel-phet** [comment]
@pixelzoom - more and more it seems Molecule Polarity will be the right one to attack next. Perhaps we should discuss at developer meeting. Seems like it would be good to coordinate this port with @jonathanolson working on Build a Molecule so that we can make nice progress on a 3d molecule viewer.

**Comment by pixelzoom** [comment]
@ariel-phet said:
> Perhaps we should discuss at developer meeting.

Label added.

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/tasks/issues/354#issuecomment-140595442, @ariel-phet proposed discussing the 3D molecule viewer in Molecule Polarity.  That is addressed in https://github.com/phetsims/molecule-polarity/issues/15, so let's move the discussion there.

**Comment by pixelzoom** [comment]
3D molecule viewer for molecule-polarity is a big(ish) project. Is there another “easy” sim port that I can work on? Met with @ariel-phet.  I suggested either rutherford-scattering or hydrogen-atom. @ariel-phet will discuss with @kathy-phet.

**Comment by ariel-phet** [comment]
Plan is to bring up at 11/12/2015 status meeting. Reassigning to @pixelzoom to bring up during his report.

**Comment by pixelzoom** [comment]
Discussed 11/12/2015. The plan is to port "Models of the Hydrogen Atom" next. The sim name won't be changing, so I can create the repo. I'll wait to start the port until @amyh-phet has made an initial "design refresh" pass.

**Comment by pixelzoom** [comment]
Work has been done on "Models of the Hydrogen Atom" (aka MOTHA), enough to jumpstart the design process.  Further work is being done as a low-priority background task, so this port is unlikely to be completed (by me, anyway) before late-2016/early-2017.  

Since I now have plenty of "port" work on my plate, unassigning this issue.  I'll reassign when I need more work.

**Comment by zepumph** [comment]
@pixelzoom, is this issue still helpful to you? If not please close. (also love that the last comment is about your current sim).

**Comment by pixelzoom** [comment]
Yes, this issue is still useful to me.  I referred to it just recently, when reporting which quantum sims could potentially be ported.
